### PR TITLE
[#79] Re Architecture SceneSequence And Several new feature

### PR DESCRIPTION
**Pull request by crexista**
**Abstruct**

1. Scene, Scenario, Sequence are redesigned, and refactored
2. Some new feature added.

**Refactor**

**A. redefine deference screen and scene**
`Screen` was just Scene's type erasure protocol
And, `Screen` had `sendTransitionRequest` and `leaveFromCurrent`.

but I redefine `Screen` and `Scene` as below.

- Screen
  1. Just Display, like a plate.
  2. Screen has two status
    - activated
    - suspended

- Scene
  1. Scene is transitionable screen.
  2. So, Scene has context and returnValue(when rewind)

So, `Screen` does not have `sendTransitionRequest` and `leaveFromCurrent` from now.
And, `Scene` has `sendTransitionRequest` and `leaveFromCurrent` insted of `Screen`.

**B. SceneSequence separate to SceneCollection, ScreenContainer**

`SceneSequence` become complex, so redesigned `SceneSequence`
So, `SceneSequence` become to extend SceneCollection and implement ScreenContainer

SceneCollection's job is stacking `Scene` and activate `Scene`
detail will be write later..

**New Feature**

- `Scene` can define associeated Type `ReturnValue`
  ReturnValue is type of argument to previous scene when rewind to back scene.

  can use like this
  ```swift
  // on Fist Scene
  sendTransitionRequest(SampleRequest(value){ (result: String) in
    print(result)
  })

  // on Second Scene
  leaveFromCurrent(returnValue: String) // SampleRequest's ResultValue Type
  ```

- `SceneSequence` can suspend.
  `SceneSequence` become to be able to suspend.
  suspended screen cannot sendTransitionRequest(if request is send, it will be ignored)

- `SceneSequenceBuilder` added.
  `SceneSequence`'s initial parameter is many.
   So `SceneSequenceBuilder` is maked.

   ```swift
   let builder = SceneSequence.builder(scene: scene, guide: rule)
                              .setup(UINavigationController(rootViewController: UIViewController()), with: false)
   sequence = builder.build(onInvoke: { (stage, scene) in
       stage.isNavigationBarHidden = true
       stage.pushViewController(scene, animated: true)
   }, onActive: { (stage, screens) in
       self.window?.rootViewController?.addChildViewController(stage)
       self.window?.rootViewController?.view.addSubview(stage.view)
   }, onSuspend: { (stage, screens) in
       stage.view.removeFromSuperview()
       stage.removeFromParentViewController()
   })

   sequence?.activate()
   ```

**Breaking Change**
- `SceneOperation`'s I/F is changed.

   **old**

   ```swift
  operation.at(Sample1BViewController.self) { (scenario) in
      scenario.given(SampleARequest.self, { Sample1AViewController() }) { (args) in
          args.stage.pushViewController(args.next, animated: true)
             return {
                args.stage.popViewController(animated: true)
             }
          }
      }
  }
  ```
  **new**

  ```swift
  operation.at(Sample1BViewController.self) { (scenario) in
      scenario.given(SampleARequest.self, transitTo: { Sample1AViewController() }) { (args) in
          args.stage.pushViewController(args.next, animated: true)
             return {
                args.stage.popViewController(animated: true)
             }
          }
      }
  }
  ```

### CODE REVIEW

File: KabuKit/Classes/core/Swift/SceneSequence.swift [hunk 1]
```diff
@@ -9,85 +9,103 @@ internal var procedureByScene = [ScreenHashWrapper : TransitionProcedure]()
...
-    
-    private var stage: StageType?
-    
-    private var guide: GuideType?
-    
-    private let dispatchQueue: DispatchQueue
+    private let firstScene: FirstScene
+    private let guide: Guide
+    private let leaveFunc: ((Guide.Stage, [Screen], ReturnValue?) -> Void)?
+    private let invokeFunc: (Guide.Stage, FirstScene) -> (() -> Void)?
```
Comment (kmhjs) [inline KabuKit/Classes/core/Swift/SceneSequence.swift]: I felt that this variable name is too ambiguous.
(And I couldn't understand the task for this variable clearly... Could you give me some explanation?)

File: KabuKit/Classes/core/Swift/Screen.swift [hunk 2]
```diff
@@ -1,72 +1,82 @@
...
-
-    func leave(_ completion: @escaping (Bool) -> Void) -> Void
+    func onActivate() -> Void
+}
 
-    func leave(_ runTransition: Bool, _ completion: @escaping (Bool) -> Void) -> Void
+public extension Screen {
+    public var isSuspended: Bool {
+        guard let behavior = screenBehaviorContainer[ScreenHashWrapper(self)] else {
+            return true
```
Comment (kmhjs) [inline KabuKit/Classes/core/Swift/Screen.swift]: Is this return value is appropriate?

File: KabuKit/Classes/core/Swift/TransitionRequest.swift [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import Foundation
+
+open class TransitionRequest<ContextType, ExpectedResultType> {
```
Comment (kmhjs) [inline KabuKit/Classes/core/Swift/TransitionRequest.swift]: Why `Expected` ?

File: KabuKit/Classes/core/Swift/SceneCollection.swift [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+     
+     設定されていない場合は登録しません
+     
+     - Parameters:
+       - newScene: 新しいScene
+       - context: 新しいSceneを起動する際に必要になるcontext
+       - callback:
+       - transition: 画面遷移ロジック
+     */
+    func stack<SceneType: Scene>(_ newScene: SceneType,
```
[the remark this answers is not in the document]
Comment (crexista) [reply KabuKit/Classes/core/Swift/SceneCollection.swift]: ok.
I'll change to what you say

File: KabuKit/Classes/core/Swift/Screen.swift (hunk 2 above)
Comment (crexista) [reply KabuKit/Classes/core/Swift/Screen.swift]: Yes.
When `Screen` is generated, `Screen`'s status is `suspend`.
So, if `ScreenBehavior` is not set, `Screen`'s status must be `suspend`.

I'll write that's comment in japanese 🤒

File: KabuKit/Classes/core/Swift/TransitionRequest.swift (hunk 3 above)
Comment (crexista) [reply KabuKit/Classes/core/Swift/TransitionRequest.swift]: The `Scene` that is received `TransitionRequest` does not necessarily reply anything.
_`Scene` can reply `nil`._
So, I named this generic type to ExpectedResultType.

But,  I think about the fact that `Scene can reply nil` is already expressed by Optional Type, now.
So, I rename `ExpectedResultType`  to `ReturnValue`
_ and rename `ContextType` to `Context`. (same to Scene)_

File: KabuKit/Classes/core/Swift/Scene.swift [hunk 5]
```diff
@@ -1,19 +1,100 @@
...
 }
 
-extension Screen where Self: Scene {
+public extension Scene {
+    
+    internal var scenario: TransitionProcedure? {
+        return procedureByScene[ScreenHashWrapper(self)]
+    }
+
+    internal var rewind: ((ReturnValue?) -> Void)? {
```
[the remark this answers is not in the document]
Comment (crexista) [reply KabuKit/Classes/core/Swift/Scene.swift]: OK
I set typealias

File: KabuKit/Classes/core/Swift/SceneSequence.swift (hunk 1 above)
Comment (crexista) [reply KabuKit/Classes/core/Swift/SceneSequence.swift]: `invokeFunc` is called at once when `SceneSequence`'s activate method is called and isStarted is false.
In a one word, this callback function supply `SceneSequence`'s initialize process.

so, maybe `initialize` fit better than invoke.

Comment (crexista) [reply KabuKit/Classes/core/Swift/SceneSequence.swift]: and now,
maybe, leaveFunc is not necessary

File: KabuKit/Classes/core/Swift/Screen.swift (hunk 2 above)
Comment (kmhjs) [reply KabuKit/Classes/core/Swift/Screen.swift]: Thanks :)

**Pull request merged**